## Model evaluation



### Why Use MAE, RMSE, and MAPE?
In forecasting model evaluation, the goal is to measure how well the model's predictions fit the actual data. To achieve this, error metrics are used to quantify the difference between predictions (`yhat`) and observed values (`y`). Three of the most common metrics for this purpose are:

#### 1. **Mean Absolute Error (MAE)**
   - **Definition**: MAE is the average of the absolute errors between the predictions and actual values. It measures the average magnitude of errors in a set of predictions, regardless of whether the errors are positive or negative.
   - **Formula**: 
     $$ \text{MAE} = \frac{1}{n} \sum_{i=1}^n |y_i - \hat{y}_i|$$
     
   - **Interpretation**: MAE provides an intuitive measure of how much our predictions deviate on average from the actual values. Since it does not square the errors, MAE is less sensitive to outliers, making it useful when we want a robust measure of accuracy that isn't heavily skewed by unexpected spikes.
   - **Applicability in COVID-19 forecasting**: If the priority is having a "typical" error that is easy to interpret in terms of daily case numbers, MAE is a good choice.

#### 2. **Root Mean Squared Error (RMSE)**
   - **Definition**: RMSE is the square root of the average of the squared differences between predicted and actual values. Unlike MAE, RMSE penalizes larger errors more heavily, which can be beneficial when large errors represent a higher risk in decision-making.
   - **Formula**:
     $$ \text{RMSE} = \sqrt{\frac{1}{n} \sum_{i=1}^n (y_i - \hat{y}_i)^2}$$
     
   - **Interpretation**: RMSE is more sensitive to outliers, which can be useful if you want to ensure that the model penalizes large errors significantly. In COVID-19 forecasting, RMSE can be particularly useful when large deviations in case numbers are concerning (for example, during peak periods).
   - **Applicability in COVID-19 forecasting**: If large deviations in case numbers are especially problematic, RMSE is a suitable metric as it "punishes" predictions that stray significantly from actual values.

#### 3. **Mean Absolute Percentage Error (MAPE)**
   - **Definition**: MAPE calculates the average of the absolute percentage errors between predictions and actual values, providing an intuitive percentage-based accuracy metric.
   - **Formula**:
     $$ \text{MAPE} = \frac{100}{n} \sum_{i=1}^n \left| \frac{y_i - \hat{y}_i}{y_i} \right|$$
     
   - **Interpretation**: MAPE expresses errors as a percentage of the actual values, making it easy to interpret in terms of relative accuracy. For instance, a MAPE of 5% means the average forecasted value is off by 5% from the actual value. However, MAPE can be sensitive to very small actual values, leading to high error rates or undefined values.
   - **Applicability in COVID-19 forecasting**: MAPE is especially useful when it's important to understand the error in relative terms, such as how much the model’s predictions deviate as a percentage of actual cases. This is beneficial for comparing performance over time or across regions with varying case numbers.

### Why Use All Three Metrics?
MAE, RMSE, and MAPE each provide unique insights into the model’s performance:

- **MAE** gives a straightforward measure of average error, useful for understanding typical deviations in absolute terms and for robust performance when outliers are present.
- **RMSE** is beneficial if we want the model to be "strict" and heavily penalize larger errors, which is valuable in situations where large deviations are particularly costly or concerning.
- **MAPE** allows us to assess the model’s accuracy in percentage terms, which is helpful for comparing accuracy across datasets of different scales or understanding performance across varying levels of COVID-19 cases.

By using all three metrics, you can achieve a more balanced and comprehensive view of the model's performance, capturing both absolute and relative error perspectives, as well as sensitivity to large deviations.

In [17]:
import numpy as np
import pandas as pd
import malib.models.metrics as mmm
import malib.data.plotting as mdp
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

In [18]:
PATH = "../../data/processed/df_clean_target_confirmed.csv"
PATH_HP_OPTUNA = "../../data/processed/prediction/optuna.csv"
PATH_HP_HYPEROPT = "../../data/processed/prediction/hyperopt.csv"
PATH_HP_PARAM_GRID = "../../data/processed/prediction/params_grid.csv"
PATH_TRAIN = "../../data/processed/train.csv"
PATH_TEST = "../../data/processed/test.csv"

In [19]:
df = pd.read_csv(PATH)
df_opt = pd.read_csv(PATH_HP_OPTUNA)
df_hyp = pd.read_csv(PATH_HP_HYPEROPT)
df_pg = pd.read_csv(PATH_HP_PARAM_GRID)
train = pd.read_csv(PATH_TRAIN)

In [20]:
list_of_models = {"optuna":df_opt,"hyperopt":df_hyp,"params_grid":df_pg}

In [21]:
rr = df_opt.shape[0]

In [22]:
for key,m in list_of_models.items():
    print(key)
    eval = mmm.daily_evaluation(df.tail(rr),m,"ds","y","yhat")
    display(eval)
    print("Metrics summary")
    pd.set_option('display.float_format', lambda x: '%.3f' % x)
    display(eval.mean(numeric_only=True).T)
    eval["y"] = eval["predict"]
    eval["ds"] = eval["Date"]
    fig = mdp.plot_interactive_forecast(df,train,eval[["ds","y"]],"ds","y")
    fig.show(renderer='iframe')

optuna


Date         true      predict        RMSE         MAE  MAPE  \
0   2020-06-20  8805336.000  8720409.730   84926.270   84926.270 0.010   
1   2020-06-21  8933875.000  8846211.760   87663.240   87663.240 0.010   
2   2020-06-22  9071733.000  8969928.960  101804.040  101804.040 0.011   
3   2020-06-23  9237071.000  9099520.160  137550.840  137550.840 0.015   
4   2020-06-24  9408254.000  9231983.250  176270.750  176270.750 0.019   
5   2020-06-25  9586141.000  9367446.120  218694.880  218694.880 0.023   
6   2020-06-26  9777487.000  9503178.150  274308.850  274308.850 0.028   
7   2020-06-27  9955597.000  9631991.550  323605.450  323605.450 0.033   
8   2020-06-28 10117227.000  9757793.580  359433.420  359433.420 0.036   
9   2020-06-29 10275799.000  9881510.780  394288.220  394288.220 0.038   
10  2020-06-30 10449697.000 10011101.980  438595.020  438595.020 0.042   
11  2020-07-01 10667386.000 10143565.070  523820.930  523820.930 0.049   
12  2020-07-02 10875091.000 10279027.940  596063.060  596063.060 0.055   
13  2020-07-03 11078585.000 10414759.970  663825.030  663825.030 0.060   
14  2020-07-04 11272152.000 10543573.370  728578.630  728578.630 0.065   
15  2020-07-05 11454847.000 10669375.400  785471.600  785471.600 0.069   
16  2020-07-06 11622190.000 10793092.600  829097.400  829097.400 0.071   
17  2020-07-07 11833034.000 10922683.800  910350.200  910350.200 0.077   
18  2020-07-08 12044836.000 11055146.890  989689.110  989689.110 0.082   
19  2020-07-09 12273063.000 11190609.760 1082453.240 1082453.240 0.088   
20  2020-07-10 12505640.000 11326341.800 1179298.200 1179298.200 0.094   
21  2020-07-11 12721968.000 11455155.190 1266812.810 1266812.810 0.100   
22  2020-07-12 12914636.000 11580957.220 1333678.780 1333678.780 0.103   
23  2020-07-13 13107415.000 11704674.430 1402740.570 1402740.570 0.107   
24  2020-07-14 13328867.000 11834265.620 1494601.380 1494601.380 0.112   
25  2020-07-15 13559984.000 11966728.710 1593255.290 1593255.290 0.117   
26  2020-07-16 13812525.000 12102191.580 1710333.420 1710333.420 0.124   
27  2020-07-17 14054563.000 12237923.620 1816639.380 1816639.380 0.129   
28  2020-07-18 14292198.000 12366737.010 1925460.990 1925460.990 0.135   
29  2020-07-19 14506845.000 12492539.040 2014305.960 2014305.960 0.139   
30  2020-07-20 14713623.000 12616256.250 2097366.750 2097366.750 0.143   
31  2020-07-21 14947078.000 12745847.440 2201230.560 2201230.560 0.147   
32  2020-07-22 15227725.000 12878310.530 2349414.470 2349414.470 0.154   
33  2020-07-23 15510481.000 13013773.400 2496707.600 2496707.600 0.161   
34  2020-07-24 15791645.000 13149505.440 2642139.560 2642139.560 0.167   
35  2020-07-25 16047190.000 13278318.830 2768871.170 2768871.170 0.173   
36  2020-07-26 16251796.000 13404120.860 2847675.140 2847675.140 0.175   
37  2020-07-27 16480485.000 13527838.070 2952646.930 2952646.930 0.179   

    Normalized RMSE  Normalized MAE  Accuracy  
0             0.010           0.010     0.990  
1             0.010           0.010     0.990  
2             0.011           0.011     0.989  
3             0.015           0.015     0.985  
4             0.019           0.019     0.981  
5             0.023           0.023     0.977  
6             0.028           0.028     0.972  
7             0.033           0.033     0.967  
8             0.036           0.036     0.964  
9             0.038           0.038     0.962  
10            0.042           0.042     0.958  
11            0.049           0.049     0.951  
12            0.055           0.055     0.945  
13            0.060           0.060     0.940  
14            0.065           0.065     0.935  
15            0.069           0.069     0.931  
16            0.071           0.071     0.929  
17            0.077           0.077     0.923  
18            0.082           0.082     0.918  
19            0.088           0.088     0.912  
20            0.094           0.094     0.906  
21            0.100           0.100     0.900  
22            0.103

Metrics summary


true              12329317.500
predict           11124063.049
RMSE               1205254.451
MAE                1205254.451
MAPE                     0.088
Normalized RMSE          0.088
Normalized MAE           0.088
Accuracy                 0.912
dtype: float64

/home/marodenas/Documentos/data_science/repositorios/covid_forecast_prophet/src/malib/data/plotting.py:361: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



hyperopt


Date         true      predict        RMSE         MAE  MAPE  \
0   2020-06-20  8805336.000  8720637.440   84698.560   84698.560 0.010   
1   2020-06-21  8933875.000  8845785.600   88089.400   88089.400 0.010   
2   2020-06-22  9071733.000  8969216.200  102516.800  102516.800 0.011   
3   2020-06-23  9237071.000  9099054.960  138016.040  138016.040 0.015   
4   2020-06-24  9408254.000  9232145.870  176108.130  176108.130 0.019   
5   2020-06-25  9586141.000  9368165.490  217975.510  217975.510 0.023   
6   2020-06-26  9777487.000  9503918.250  273568.750  273568.750 0.028   
7   2020-06-27  9955597.000  9632316.920  323280.080  323280.080 0.032   
8   2020-06-28 10117227.000  9757465.090  359761.910  359761.910 0.036   
9   2020-06-29 10275799.000  9880895.690  394903.310  394903.310 0.038   
10  2020-06-30 10449697.000 10010734.440  438962.560  438962.560 0.042   
11  2020-07-01 10667386.000 10143825.350  523560.650  523560.650 0.049   
12  2020-07-02 10875091.000 10279844.970  595246.030  595246.030 0.055   
13  2020-07-03 11078585.000 10415597.740  662987.260  662987.260 0.060   
14  2020-07-04 11272152.000 10543996.410  728155.590  728155.590 0.065   
15  2020-07-05 11454847.000 10669144.570  785702.430  785702.430 0.069   
16  2020-07-06 11622190.000 10792575.180  829614.820  829614.820 0.071   
17  2020-07-07 11833034.000 10922413.930  910620.070  910620.070 0.077   
18  2020-07-08 12044836.000 11055504.840  989331.160  989331.160 0.082   
19  2020-07-09 12273063.000 11191524.460 1081538.540 1081538.540 0.088   
20  2020-07-10 12505640.000 11327277.220 1178362.780 1178362.780 0.094   
21  2020-07-11 12721968.000 11455675.900 1266292.100 1266292.100 0.100   
22  2020-07-12 12914636.000 11580824.060 1333811.940 1333811.940 0.103   
23  2020-07-13 13107415.000 11704254.660 1403160.340 1403160.340 0.107   
24  2020-07-14 13328867.000 11834093.410 1494773.590 1494773.590 0.112   
25  2020-07-15 13559984.000 11967184.330 1592799.670 1592799.670 0.117   
26  2020-07-16 13812525.000 12103203.950 1709321.050 1709321.050 0.124   
27  2020-07-17 14054563.000 12238956.710 1815606.290 1815606.290 0.129   
28  2020-07-18 14292198.000 12367355.380 1924842.620 1924842.620 0.135   
29  2020-07-19 14506845.000 12492503.540 2014341.460 2014341.460 0.139   
30  2020-07-20 14713623.000 12615934.150 2097688.850 2097688.850 0.143   
31  2020-07-21 14947078.000 12745772.900 2201305.100 2201305.100 0.147   
32  2020-07-22 15227725.000 12878863.810 2348861.190 2348861.190 0.154   
33  2020-07-23 15510481.000 13014883.430 2495597.570 2495597.570 0.161   
34  2020-07-24 15791645.000 13150636.200 2641008.800 2641008.800 0.167   
35  2020-07-25 16047190.000 13279034.870 2768155.130 2768155.130 0.173   
36  2020-07-26 16251796.000 13404183.030 2847612.970 2847612.970 0.175   
37  2020-07-27 16480485.000 13527613.630 2952871.370 2952871.370 0.179   

    Normalized RMSE  Normalized MAE  Accuracy  
0             0.010           0.010     0.990  
1             0.010           0.010     0.990  
2             0.011           0.011     0.989  
3             0.015           0.015     0.985  
4             0.019           0.019     0.981  
5             0.023           0.023     0.977  
6             0.028           0.028     0.972  
7             0.032           0.032     0.968  
8             0.036           0.036     0.964  
9             0.038           0.038     0.962  
10            0.042           0.042     0.958  
11            0.049           0.049     0.951  
12            0.055           0.055     0.945  
13            0.060           0.060     0.940  
14            0.065           0.065     0.935  
15            0.069           0.069     0.931  
16            0.071           0.071     0.929  
17            0.077           0.077     0.923  
18            0.082           0.082     0.918  
19            0.088           0.088     0.912  
20            0.094           0.094     0.906  
21            0.100           0.100     0.900  
22            0.103

Metrics summary


true              12329317.500
predict           11124289.857
RMSE               1205027.643
MAE                1205027.643
MAPE                     0.088
Normalized RMSE          0.088
Normalized MAE           0.088
Accuracy                 0.912
dtype: float64

/home/marodenas/Documentos/data_science/repositorios/covid_forecast_prophet/src/malib/data/plotting.py:361: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



params_grid


Date         true      predict        RMSE         MAE  MAPE  \
0   2020-06-20  8805336.000  8724912.110   80423.890   80423.890 0.009   
1   2020-06-21  8933875.000  8850947.000   82928.000   82928.000 0.009   
2   2020-06-22  9071733.000  8974638.760   97094.240   97094.240 0.011   
3   2020-06-23  9237071.000  9103851.900  133219.100  133219.100 0.014   
4   2020-06-24  9408254.000  9236767.220  171486.780  171486.780 0.018   
5   2020-06-25  9586141.000  9374286.740  211854.260  211854.260 0.022   
6   2020-06-26  9777487.000  9510699.880  266787.120  266787.120 0.027   
7   2020-06-27  9955597.000  9640593.220  315003.780  315003.780 0.032   
8   2020-06-28 10117227.000  9766628.110  350598.890  350598.890 0.035   
9   2020-06-29 10275799.000  9890319.870  385479.130  385479.130 0.038   
10  2020-06-30 10449697.000 10019533.010  430163.990  430163.990 0.041   
11  2020-07-01 10667386.000 10152448.330  514937.670  514937.670 0.048   
12  2020-07-02 10875091.000 10289967.850  585123.150  585123.150 0.054   
13  2020-07-03 11078585.000 10426380.990  652204.010  652204.010 0.059   
14  2020-07-04 11272152.000 10556274.330  715877.670  715877.670 0.064   
15  2020-07-05 11454847.000 10682309.220  772537.780  772537.780 0.067   
16  2020-07-06 11622190.000 10806000.980  816189.020  816189.020 0.070   
17  2020-07-07 11833034.000 10935214.120  897819.880  897819.880 0.076   
18  2020-07-08 12044836.000 11068129.440  976706.560  976706.560 0.081   
19  2020-07-09 12273063.000 11205648.960 1067414.040 1067414.040 0.087   
20  2020-07-10 12505640.000 11342062.100 1163577.900 1163577.900 0.093   
21  2020-07-11 12721968.000 11471955.440 1250012.560 1250012.560 0.098   
22  2020-07-12 12914636.000 11597990.330 1316645.670 1316645.670 0.102   
23  2020-07-13 13107415.000 11721682.090 1385732.910 1385732.910 0.106   
24  2020-07-14 13328867.000 11850895.220 1477971.780 1477971.780 0.111   
25  2020-07-15 13559984.000 11983810.540 1576173.460 1576173.460 0.116   
26  2020-07-16 13812525.000 12121330.070 1691194.930 1691194.930 0.122   
27  2020-07-17 14054563.000 12257743.210 1796819.790 1796819.790 0.128   
28  2020-07-18 14292198.000 12387636.550 1904561.450 1904561.450 0.133   
29  2020-07-19 14506845.000 12513671.430 1993173.570 1993173.570 0.137   
30  2020-07-20 14713623.000 12637363.200 2076259.800 2076259.800 0.141   
31  2020-07-21 14947078.000 12766576.330 2180501.670 2180501.670 0.146   
32  2020-07-22 15227725.000 12899491.650 2328233.350 2328233.350 0.153   
33  2020-07-23 15510481.000 13037011.180 2473469.820 2473469.820 0.159   
34  2020-07-24 15791645.000 13173424.310 2618220.690 2618220.690 0.166   
35  2020-07-25 16047190.000 13303317.650 2743872.350 2743872.350 0.171   
36  2020-07-26 16251796.000 13429352.540 2822443.460 2822443.460 0.174   
37  2020-07-27 16480485.000 13553044.310 2927440.690 2927440.690 0.178   

    Normalized RMSE  Normalized MAE  Accuracy  
0             0.009           0.009     0.991  
1             0.009           0.009     0.991  
2             0.011           0.011     0.989  
3             0.014           0.014     0.986  
4             0.018           0.018     0.982  
5             0.022           0.022     0.978  
6             0.027           0.027     0.973  
7             0.032           0.032     0.968  
8             0.035           0.035     0.965  
9             0.038           0.038     0.962  
10            0.041           0.041     0.959  
11            0.048           0.048     0.952  
12            0.054           0.054     0.946  
13            0.059           0.059     0.941  
14            0.064           0.064     0.936  
15            0.067           0.067     0.933  
16            0.070           0.070     0.930  
17            0.076           0.076     0.924  
18            0.081           0.081     0.919  
19            0.087           0.087     0.913  
20            0.093           0.093     0.907  
21            0.098           0.098     0.902  
22            0.102

Metrics summary


true              12329317.500
predict           11138523.952
RMSE               1190793.548
MAE                1190793.548
MAPE                     0.087
Normalized RMSE          0.087
Normalized MAE           0.087
Accuracy                 0.913
dtype: float64

/home/marodenas/Documentos/data_science/repositorios/covid_forecast_prophet/src/malib/data/plotting.py:361: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



### Conclusions

#### 1. Choice of Hyperparameter Tuning Method: Optuna, Param Grid, or Hyperopt

When comparing the performance metrics across Optuna, Param Grid, and Hyperopt, we observe the following:

- **Accuracy**: All three methods achieve a very similar average accuracy of approximately 0.912–0.913. This suggests that each tuning method is producing models with a comparable level of overall fit. However, **Param Grid** has a slightly higher accuracy (0.913) than the others, though the difference is minimal.
  
- **RMSE and MAE**: Param Grid achieves the lowest overall RMSE (1,189,803) and MAE (1,189,803) among the three methods. Hyperopt follows closely, while Optuna has a slightly higher RMSE and MAE. This indicates that **Param Grid** provides the best absolute fit in terms of error, as both RMSE and MAE are minimized with this method.

- **MAPE and Normalized RMSE/MAE**: The average MAPE and normalized RMSE/MAE for all three methods are also very similar (around 0.087–0.088). However, **Param Grid** has a marginally better normalized RMSE and MAE (0.087) compared to the other methods (0.088 for Optuna and Hyperopt).

**Recommendation**: Based on these metrics, **Param Grid** is the preferred hyperparameter tuning method. It provides the lowest error (RMSE and MAE) and slightly higher accuracy, indicating a marginally better fit. This advantage, though subtle, makes Param Grid the most effective method for hyperparameter tuning in this case.

---

#### 2. Evaluation of Prophet’s Forecasting Performance

The evaluation metrics give insights into the quality of the Prophet model’s forecast:

- **Overall Fit**: With a **MAPE around 8.7%–8.8%**, the Prophet model shows reasonable accuracy, though it suggests there is still some level of deviation from the actual data. A MAPE below 10% is often considered a sign of a fairly good fit in time series forecasting, indicating that Prophet is generally capturing the trend.

- **RMSE and MAE Values**: The RMSE values around 1,200,000 indicate an average forecast deviation in the range of 1.2 million units per prediction. Although these errors are sizable, they may be expected in high-scale time series data (like COVID-19 case numbers) where the absolute values are high. The **normalized RMSE and MAE (~8.7%)** imply that the error is roughly 8.7% of the mean true values, which provides a clearer interpretation of the model’s performance relative to the data scale.

- **Accuracy**: The accuracy values close to 91.3% across the models indicate that Prophet provides a consistent fit across different days, achieving reliable performance over the forecast horizon. However, the daily metrics reveal that the error increases slightly as the forecast horizon extends, which is typical in time series forecasting due to the cumulative uncertainty in the data.

**Overall Assessment**: The Prophet model is performing reasonably well in capturing the overall trend, achieving an average error (MAPE) of around 8.7% and an accuracy close to 91.3%. This indicates a fairly good fit to the observed data. However, there is room for improvement if higher accuracy is required, particularly for reducing absolute error (RMSE) on larger-scale predictions. Further fine-tuning or considering additional seasonal adjustments or other modeling approaches could potentially enhance forecast accuracy for more precise applications.